File: server-shared/src/test/java/eu/darken/porter/core/PorterCoreAttachTest.java
package eu.darken.porter.core;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static rikka.shizuku.server.ServerTestSupport.application;
import static rikka.shizuku.server.ServerTestSupport.entry;
import static rikka.shizuku.server.ServerTestSupport.newCore;

import android.os.Handler;
import android.os.IBinder;
import android.os.RemoteException;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import eu.darken.porter.endpoint.PorterClientCallback;
import eu.darken.porter.server.IPorterApplication;
import rikka.shizuku.server.ClientManager;
import rikka.shizuku.server.ClientRecord;
import rikka.shizuku.server.ConfigManager;
import rikka.shizuku.server.ServerTestSupport.TestPolicy;
import rikka.shizuku.server.ServerTestSupport.TestUserServiceManager;
import rikka.shizuku.server.UserServiceManager;
import rikka.shizuku.server.legacy.LegacyClientCallback;
import rikka.shizuku.server.util.HandlerUtil;

/** Which process gets a record, through which endpoint, and what the policy sees on the way. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class PorterCoreAttachTest {

    private static final String PACKAGE = "eu.darken.porter.probe";
    private static final String OTHER_PACKAGE = "eu.darken.porter.probe.other";
    /** User 10, app id 10200, so the user id derived from it is not the trivial zero. */
    private static final int CLIENT_UID = 1010200;
    private static final int CLIENT_PID = 45678;
    private static final int API_LEVEL = 13;

    private ConfigManager config;
    private ClientManager<ConfigManager> clients;
    private RecordingPolicy policy;
    private PorterCore<UserServiceManager, ClientManager<ConfigManager>, ConfigManager> core;
    private CallerIdentity caller;

    /** Whether a record was already there each time the policy was told an attach is starting. */
    private static class RecordingPolicy extends TestPolicy {

        private final ClientManager<ConfigManager> clients;
        final List<Boolean> recordExistedWhenAttaching = new ArrayList<>();

        RecordingPolicy(ClientManager<ConfigManager> clients) {
            this.clients = clients;
        }

        @Override
        public void onAttaching(CallerIdentity caller, String packageName) {
            super.onAttaching(caller, packageName);
            recordExistedWhenAttaching.add(clients.findClient(caller.uid, caller.pid) != null);
        }
    }

    @Before
    public void setup() {
        HandlerUtil.setMainHandler(mock(Handler.class));
        config = mock(ConfigManager.class);
        clients = new ClientManager<>(config);
        policy = new RecordingPolicy(clients);
        core = newCore(clients, new TestUserServiceManager(), config, policy,
                uid -> Arrays.asList(PACKAGE, OTHER_PACKAGE));

        caller = new CallerIdentity(CLIENT_UID, CLIENT_PID);
    }

    private static ClientCallback legacyCallback() {
        return new LegacyClientCallback(application(mock(IBinder.class)));
    }

    private static ClientCallback porterCallback() {
        IPorterApplication application = mock(IPorterApplication.class);
        when(application.asBinder()).thenReturn(mock(IBinder.class));
        return new PorterClientCallback(application);
    }

    @Test
    public void aProcessAttachedThroughShizukuCannotAttachThroughPorter() {
        core.attach(caller, PACKAGE, legacyCallback(), API_LEVEL);

        IllegalStateException e = assertThrows(IllegalStateException.class,
                () -> core.attach(caller, PACKAGE, porterCallback(), API_LEVEL));

        assertTrue(e.getMessage(), e.getMessage().contains("is attached through another endpoint"));
        assertEquals(1, clients.findClients(CLIENT_UID).size());
    }

    @Test
    public void aProcessAttachedThroughPorterCannotAttachThroughShizuku() {
        core.attach(caller, PACKAGE, porterCallback(), API_LEVEL);

        IllegalStateException e = assertThrows(IllegalStateException.class,
                () -> core.attach(caller, PACKAGE, legacyCallback(), API_LEVEL));

        assertTrue(e.getMessage(), e.getMessage().contains("is attached through another endpoint"));
        assertEquals(1, clients.findClients(CLIENT_UID).size());
    }

    @Test
    public void aReattachMayNotRenameTheClient() {
        core.attach(caller, PACKAGE, porterCallback(), API_LEVEL);

        assertThrows(SecurityException.class,
                () -> core.attach(caller, OTHER_PACKAGE, porterCallback(), API_LEVEL));

        assertEquals(1, clients.findClients(CLIENT_UID).size());
        assertEquals(PACKAGE, clients.findClient(CLIENT_UID, CLIENT_PID).packageName);
    }

    @Test
    public void theSecondAttachFromAProcessReusesTheRecordItFinds() {
        AttachResult first = core.attach(caller, PACKAGE, porterCallback(), API_LEVEL);
        AttachResult second = core.attach(caller, PACKAGE, porterCallback(), API_LEVEL);

        assertTrue(first.created);
        assertFalse(second.created);
        assertSame(first.record, second.record);

        assertEquals(Arrays.asList(PACKAGE, PACKAGE), policy.attachingPackages);
        assertEquals("the first attach is told before the record exists",
                Arrays.asList(false, true), policy.recordExistedWhenAttaching);
    }

    @Test
    public void aDeadCallbackBinderIsReportedAsIllegalState() throws Exception {
        IBinder dead = mock(IBinder.class);
        doThrow(new RemoteException()).when(dead).linkToDeath(any(IBinder.DeathRecipient.class), anyInt());
        IPorterApplication application = mock(IPorterApplication.class);
        when(application.asBinder()).thenReturn(dead);

        assertThrows(IllegalStateException.class,
                () -> core.attach(caller, PACKAGE, new PorterClientCallback(application), API_LEVEL));

        assertNull(clients.findClient(CLIENT_UID, CLIENT_PID));
    }

    @Test
    public void aPackageThatDoesNotBelongToTheCallerIsRefusedBeforeThePolicyHears() {
        PorterCore<UserServiceManager, ClientManager<ConfigManager>, ConfigManager> foreign =
                newCore(clients, new TestUserServiceManager(), config, policy,
                        uid -> Collections.singletonList(OTHER_PACKAGE));

        assertThrows(SecurityException.class,
                () -> foreign.attach(caller, PACKAGE, porterCallback(), API_LEVEL));

        assertNull(clients.findClient(CLIENT_UID, CLIENT_PID));
        assertTrue(policy.attachingPackages.isEmpty());
    }

    @Test
    public void anAttachedRecordCarriesWhatTheCallerWasGranted() {
        when(config.find(CLIENT_UID)).thenReturn(entry(true, false));

        ClientRecord record = core.attach(caller, PACKAGE, porterCallback(), API_LEVEL).record;

        assertTrue(record.allowed);
        assertEquals(PACKAGE, record.packageName);
        assertEquals(API_LEVEL, record.apiVersion);
    }
}

File: server-shared/src/test/java/eu/darken/porter/core/PorterCoreAttachTest.kt
package eu.darken.porter.core

import android.os.Handler
import android.os.IBinder
import android.os.RemoteException
import eu.darken.porter.endpoint.PorterClientCallback
import eu.darken.porter.server.IPorterApplication
import java.util.function.IntFunction
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertSame
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mockito.doThrow
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import rikka.shizuku.server.ClientManager
import rikka.shizuku.server.ConfigManager
import rikka.shizuku.server.ServerTestSupport
import rikka.shizuku.server.UserServiceManager
import rikka.shizuku.server.legacy.LegacyClientCallback
import rikka.shizuku.server.util.HandlerUtil

/** Which process gets a record, through which endpoint, and what the policy sees on the way. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
class PorterCoreAttachTest {

    private lateinit var config: ConfigManager
    private lateinit var clients: ClientManager<ConfigManager>
    private lateinit var policy: RecordingPolicy
    private lateinit var core: PorterCore<UserServiceManager, ClientManager<ConfigManager>, ConfigManager>
    private lateinit var caller: CallerIdentity

    /** Whether a record was already there each time the policy was told an attach is starting. */
    private class RecordingPolicy(private val clients: ClientManager<ConfigManager>) : ServerTestSupport.TestPolicy() {

        val recordExistedWhenAttaching = ArrayList<Boolean>()

        override fun onAttaching(caller: CallerIdentity, packageName: String) {
            super.onAttaching(caller, packageName)
            recordExistedWhenAttaching.add(clients.findClient(caller.uid, caller.pid) != null)
        }
    }

    @Before
    fun setup() {
        HandlerUtil.mainHandler = mock(Handler::class.java)
        config = mock(ConfigManager::class.java)
        clients = ClientManager(config)
        policy = RecordingPolicy(clients)
        core = ServerTestSupport.newCore(clients, ServerTestSupport.TestUserServiceManager(), config, policy) {
            listOf(PACKAGE, OTHER_PACKAGE)
        }

        caller = CallerIdentity(CLIENT_UID, CLIENT_PID)
    }

    @Test
    fun aProcessAttachedThroughShizukuCannotAttachThroughPorter() {
        core.attach(caller, PACKAGE, legacyCallback(), API_LEVEL)

        val e = assertThrows(IllegalStateException::class.java) {
            core.attach(caller, PACKAGE, porterCallback(), API_LEVEL)
        }

        assertTrue(e.message, e.message!!.contains("is attached through another endpoint"))
        assertEquals(1, clients.findClients(CLIENT_UID).size)
    }

    @Test
    fun aProcessAttachedThroughPorterCannotAttachThroughShizuku() {
        core.attach(caller, PACKAGE, porterCallback(), API_LEVEL)

        val e = assertThrows(IllegalStateException::class.java) {
            core.attach(caller, PACKAGE, legacyCallback(), API_LEVEL)
        }

        assertTrue(e.message, e.message!!.contains("is attached through another endpoint"))
        assertEquals(1, clients.findClients(CLIENT_UID).size)
    }

    @Test
    fun aReattachMayNotRenameTheClient() {
        core.attach(caller, PACKAGE, porterCallback(), API_LEVEL)

        assertThrows(SecurityException::class.java) {
            core.attach(caller, OTHER_PACKAGE, porterCallback(), API_LEVEL)
        }

        assertEquals(1, clients.findClients(CLIENT_UID).size)
        assertEquals(PACKAGE, clients.findClient(CLIENT_UID, CLIENT_PID)!!.packageName)
    }

    @Test
    fun theSecondAttachFromAProcessReusesTheRecordItFinds() {
        val first = core.attach(caller, PACKAGE, porterCallback(), API_LEVEL)
        val second = core.attach(caller, PACKAGE, porterCallback(), API_LEVEL)

        assertTrue(first.created)
        assertFalse(second.created)
        assertSame(first.record, second.record)

        assertEquals(listOf(PACKAGE, PACKAGE), policy.attachingPackages)
        assertEquals(
            "the first attach is told before the record exists",
            listOf(false, true), policy.recordExistedWhenAttaching,
        )
    }

    @Test
    @Throws(Exception::class)
    fun aDeadCallbackBinderIsReportedAsIllegalState() {
        val dead = mock(IBinder::class.java)
        doThrow(RemoteException()).`when`(dead).linkToDeath(any(IBinder.DeathRecipient::class.java), anyInt())
        val application = mock(IPorterApplication::class.java)
        `when`(application.asBinder()).thenReturn(dead)

        assertThrows(IllegalStateException::class.java) {
            core.attach(caller, PACKAGE, PorterClientCallback(application), API_LEVEL)
        }

        assertNull(clients.findClient(CLIENT_UID, CLIENT_PID))
    }

    @Test
    fun aPackageThatDoesNotBelongToTheCallerIsRefusedBeforeThePolicyHears() {
        val foreign = ServerTestSupport.newCore(
            clients, ServerTestSupport.TestUserServiceManager(), config, policy,
            IntFunction { listOf(OTHER_PACKAGE) },
        )

        assertThrows(SecurityException::class.java) {
            foreign.attach(caller, PACKAGE, porterCallback(), API_LEVEL)
        }

        assertNull(clients.findClient(CLIENT_UID, CLIENT_PID))
        assertTrue(policy.attachingPackages.isEmpty())
    }

    @Test
    fun anAttachedRecordCarriesWhatTheCallerWasGranted() {
        `when`(config.find(CLIENT_UID)).thenReturn(ServerTestSupport.entry(true, false))

        val record = core.attach(caller, PACKAGE, porterCallback(), API_LEVEL).record

        assertTrue(record.allowed)
        assertEquals(PACKAGE, record.packageName)
        assertEquals(API_LEVEL, record.apiVersion)
    }

    private companion object {
        const val PACKAGE = "eu.darken.porter.probe"
        const val OTHER_PACKAGE = "eu.darken.porter.probe.other"

        /** User 10, app id 10200, so the user id derived from it is not the trivial zero. */
        const val CLIENT_UID = 1010200
        const val CLIENT_PID = 45678
        const val API_LEVEL = 13

        fun legacyCallback(): ClientCallback =
            LegacyClientCallback(ServerTestSupport.application(mock(IBinder::class.java)))

        fun porterCallback(): ClientCallback {
            val application = mock(IPorterApplication::class.java)
            `when`(application.asBinder()).thenReturn(mock(IBinder::class.java))
            return PorterClientCallback(application)
        }
    }
}
